File: Java/프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    Map<Integer, List<Integer>> map = new HashMap<>();
    boolean[] visited;
    int count;
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        for (int[] wire : wires) {
            map.putIfAbsent(wire[0], new ArrayList<>());
            map.get(wire[0]).add(wire[1]);
            map.putIfAbsent(wire[1], new ArrayList<>());
            map.get(wire[1]).add(wire[0]);
        }

        for (int[] wire : wires) {
            count = 0;
            visited = new boolean[n + 1];
            map.get(wire[0]).remove((Integer) wire[1]);
            map.get(wire[1]).remove((Integer) wire[0]);
            dfs(wire[1]);
            answer = Math.min(answer, Math.abs(count - (n - count)));
            map.get(wire[0]).add(wire[1]);
            map.get(wire[1]).add(wire[0]);
        }
        return answer;
    }

    private void dfs(int i) {
        visited[i] = true;
        count++;
        for (int j : map.get(i)) {
            if (!visited[j]) {
                dfs(j);
            }
        }
    }
}

File: Python/프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import defaultdict


def solution(n, wires):
    answer = float("inf")
    d = defaultdict(list)

    for a, b in wires:
        d[a].append(b)
        d[b].append(a)

    def dfs(num):
        visited[num] = True
        count = 1

        for i in d[num]:
            if not visited[i]:
                count += dfs(i)

        return count

    for a, b in wires:
        visited = [False] * (n + 1)

        d[a].remove(b)
        d[b].remove(a)

        count = dfs(b)
        answer = min(answer, abs(count - (n - count)))

        d[a].append(b)
        d[b].append(a)

    return answer